// node-infinito-wallet/lib/messages.js
module.exports = {
  internal_error: {
    code: 'infinito.wallet.internal_error',
    message: 'Internal error'
  },
  missing_parameter: {
    code: 'infinito.wallet.missing_parameter',
    message: 'Missing required parameter %s'
  },
  invalid_cointype: {
    code: 'infinito.wallet.invalid_cointype',
    message: 'Invalid coin type %s'
  },
  invalid_account: {
    code: 'infinito.wallet.invalid_account',
    message: 'Invalid account %s'
  },
  invalid_version: {
    code: 'infinito.wallet.invalid_version',
    message: 'Invalid version %s'
  },
  incorrect_cointype: {
    code: 'infinito.wallet.incorrect_cointype',
    message: 'coin type is inccorrect'
  },
  create_crawtx_fail: {
    code: 'infinito.wallet.create_crawtx_fail',
    message: 'Fail when create crawtx'
  },
  invalid_type: {
    code: 'infinito.wallet.invalid_type',
    message: 'Invalid type %s'
  },
  create_transaction_fail: {
    code: 'infinito.wallet.create_transaction_fail',
    message: 'Fail when create transaction'
  },
};
